### PR TITLE
Add AJDC and an example on artifact correction, 2nd try

### PR DESCRIPTION
**Pull request by qbarthelemy**
This PR:

- adds a class `AJDC` in `spatialfilters`, useful for SOS-based BSS and group BSS,

- adds an example in the gallery, on EEG artifact correction by AJDC, and a comparison with HOS-BSS (ICA-Infomax).

Discussions about implementation can be found in #90.

### CODE REVIEW

**Comment by lkorczowski** [comment]
@agramfort We would like to present the last updates of pyriemann during BCI Meeting W13. A demo will be presented to show that is it easily compatible with MNE. Do you think that this PR can be merge before Wednesday?

Can I help?

https://bcisociety.org/workshops/#session4

File: pyriemann/spatialfilters.py [hunk 1]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+
+
+class AJDC(BaseEstimator, TransformerMixin):
+    """Implementation of the AJDC.
+
+    The approximate joint diagonalization of Fourier cospectral matrices (AJDC)
+    [1] is a versatile tool for blind source separation (BSS) tasks based on
+    Second-Order Statistics (SOS), estimating spectrally uncorrelated sources.
+
+    It can be apply:
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:481]:
```suggestion
    It can be applied:
```

File: pyriemann/spatialfilters.py [hunk 2]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        The percentage of overlap between window.
+    fmin : float | None , (default None)
+        The minimal frequency to be returned. Since BSS models assume zero-mean
+        processes, the first cospectrum (0 Hz) must be excluded.
+    fmax : float | None , (default None)
+        The maximal frequency to be returned.
+    fs : float | None , (default None)
+        The sampling frequency of the signal.
+    expl_var : float (default 0.999)
+        The percentage of explained variance in ]0, 1], for dimension reduction
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:508]:
```suggestion
        The percentage of explained variance in (0, 1], for dimension reduction
```
let's us English notations https://en.wikipedia.org/wiki/Interval_(mathematics)

File: pyriemann/spatialfilters.py [hunk 3]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+    [1] M. Congedo, C. Gouy-Pailler, C. Jutten, "On the blind source separation
+    of human electroencephalogram by approximate joint diagonalization of
+    second order statistics", Clin Neurophysiol, 2008
+
+    [2] M. Congedo, R. John, D. De Ridder, L. Prichep, "Group indepedent
+    component analysis of resting state EEG in large normative samples",
+    Int J Psychophysiol, 2010
+
+    [3] D.-T. Pham, "Joint approximate diagonalization of positive definite
+    Hermitian matrices", SIAM J Matrix Anal Appl, 2001
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:547]: can you use references with url links as requested in https://github.com/pyRiemann/pyRiemann/pull/113/files ?

File: pyriemann/spatialfilters.py [hunk 4]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+    """
+
+    def __init__(self, window=128, overlap=0.5, fmin=None, fmax=None, fs=None,
+                 expl_var=0.999, verbose=True):
+        """Init."""
+        if not 0 < expl_var <= 1:
+            raise ValueError('Parameter expl_var must be included in ]0, 1]')
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:554]:
```suggestion
            raise ValueError('Parameter expl_var must be included in (0, 1]')
```

File: pyriemann/spatialfilters.py [hunk 5]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        return self
+
+    def transform(self, X):
+        """Transform channel space to source space, applying forward spatial
+        filters.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials in channel space.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:658]:
```suggestion
            Trials in channel space.
```

File: pyriemann/spatialfilters.py [hunk 6]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+
+        Returns
+        -------
+        source : ndarray, shape (n_trials, n_sources, n_samples)
+            ndarray of trials in source space.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:663]:
```suggestion
            Trials in source space.
```

File: pyriemann/spatialfilters.py [hunk 7]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+
+    def fit(self, X, y=None):
+        """Fit.
+
+        Compute and diagonalize cospectra, to estimate forward and backward
+        spatial filters.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_subjects, n_conditions, n_channels, n_samples) | list of n_subjects of list of n_conditions ndarray of shape (n_channels, n_samples), with same n_conditions and n_channels but different n_samples
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:572]: not sure what is the convention in pyRiemann but I would use n_times and not n_samples here. n_samples is confusing with sklearn naming

File: pyriemann/spatialfilters.py [hunk 8]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+
+    def inverse_transform(self, X, supp=None):
+        """Transform source space to channel space, applying backward spatial
+        filters, with the possibility to suppress some sources, like in BSS
+        filtering/denoising.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_sources, n_samples)
+            ndarray of trials in source space.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:681]:
```suggestion
            Trials in source space.
```

File: pyriemann/spatialfilters.py [hunk 9]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        supp : list of int | None, (default None)
+            List of indices of sources to suppress.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:683]:
```suggestion
            Indices of sources to suppress.
```
if None?

File: pyriemann/spatialfilters.py [hunk 10]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+
+        Returns
+        -------
+        signal : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials in channel space.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:688]:
```suggestion
            Trials in channel space.
```

File: pyriemann/spatialfilters.py [hunk 11]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        """
+        if X.ndim != 3:
+            raise ValueError('X must have 3 dimensions')
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:691]:
```suggestion
            raise ValueError('X must have 3 dimensions (Got %d)' % X.ndim)
```

File: pyriemann/spatialfilters.py [hunk 12]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        if X.shape[1] != self.n_sources_:
+            raise ValueError('X has not the good number of sources')
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:693]:
```suggestion
            raise ValueError('X does not have the good number of sources. Should be %d but got %d' % ...)
```

File: pyriemann/spatialfilters.py [hunk 13]
```diff
@@ -467,3 +469,266 @@ def fit(self, X, y):
...
+        signal = self.backward_filters_ @ denois @ X
+        return signal
+
+    def get_src_expl_var(self, X):
+        """Estimate explained variances of sources, Appendix D in [1].
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_trials, n_channels, n_samples)
+            ndarray of trials in channel space.
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:713]:
```suggestion
            Trials in channel space.
```
basically don't repeat the type when it's written above

File: pyriemann/utils/viz.py [hunk 14]
```diff
@@ -46,3 +46,33 @@ def plot_embedding(X, y=None, metric='riemann',
...
+
+def plot_cospectra(cosp, freqs, ylabels=None, title='Cospectra'):
+    """Plot cospectral matrices
+
+    Parameters
+    ----------
+    cosp : ndarray, shape (n_freqs, n_channels, n_channels)
+        ndarray of cospectra.
+    freqs : ndarray, shape (n_freqs,)
+        The frequencies associated to cospectra.
```
Comment (agramfort) [inline pyriemann/utils/viz.py:59]: Returns
missing

File: examples/artifacts/plot_correct_ajdc_EEG.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    with open(fname, "r") as f:
+        content = f.readline().split()
+        return content[:-1], int(content[-1])
+
+
+###############################################################################
+# Load EEG data
+# -------------
+
+fname = '../data/sample-blinks.txt'
```
Comment (agramfort) [inline examples/artifacts/plot_correct_ajdc_EEG.py:41]: pypi limits source distributions and it's a good idea to keep the soruce repo as small as possible.
Can you avoid using a .txt file for this reason? Maybe .npy or .txt.gz ? numpy will have no
problem with .txt.gz

File: pyriemann/spatialfilters.py (hunk 3 above)
Comment (qbarthelemy) [reply pyriemann/spatialfilters.py:547]: Ok, but I don't know why it creates a line break in the doc.

File: pyriemann/spatialfilters.py (hunk 7 above)
Comment (qbarthelemy) [reply pyriemann/spatialfilters.py:572]: In my PR, I have simply re-used the dimension names of the Covariances class: https://github.com/pyRiemann/pyRiemann/blob/master/pyriemann/estimation.py#L67

But I do agree that _samples_ is an ambiguous term, with different interpretations depending the context. What about another issue to homogenize dimension names in pyRiemann (partially related to #25)?

**Pull request merged**